src/game/common/rts/science.cpp

```diff
@@ -55,16 +74,139 @@ void ScienceStore::Reset()
 
-    for (auto it = m_infoVec.begin(); it != m_infoVec.end(); ++it) {
-        ScienceInfo *info = (ScienceInfo *)((*it)->Get_Final_Override());
+    captainslog_relassert(
+        Has_Science(key), 0xDEAD0006, "Science name %s not known! (Did you define it in science.ini?)", name);
+    return key;
+}
+
+// zh: 0x004887F0 wb: 0x0072727E
+ScienceType ScienceStore::Get_Science_From_Internal_Name(const Utf8String &name) const
+{
+    if (name.Is_Empty()) {
+        return ScienceType::SCIENCE_INVALID;
+    }
+    return static_cast<ScienceType>(g_theNameKeyGenerator->Name_To_Key(name.Str()));
+}
 
+// zh: 0x00488820 wb: 0x007272BE
+Utf8String ScienceStore::Get_Internal_Name_From_Science(ScienceType science) const
+{
+    if (science == ScienceType::SCIENCE_INVALID) {
+        return Utf8String::s_emptyString;
+    }
+    return g_theNameKeyGenerator->Key_To_Name(static_cast<NameKeyType>(science));
+}
+
+// wb 0x0072731A
+std::vector<Utf8String> ScienceStore::Get_All_Science() const
+{
+    std::vector<Utf8String> all_science;
+    for (const auto *science : m_infoVec) {
+        const auto key = static_cast<const ScienceInfo *>(science->Get_Final_Override())->Get_Name_Key();
+        all_science.push_back(g_theNameKeyGenerator->Key_To_Name(static_cast<NameKeyType>(key)));
+    }
+    return all_science;
+}
+
+// zh: 0x00488C40 wb: 0x00727753
+int32_t ScienceStore::Get_Science_Purchase_Cost(ScienceType science) const
+{
+    auto *info = Get_Science_Info(science);
+    return info != nullptr ? info->Get_Purchase_Cost() : 0;
+}
+
+// zh: 0x00488C90 wb: 0x00727781
+bool ScienceStore::Is_Science_Grantable(ScienceType science) const
+{
+    auto *info = Get_Science_Info(science);
+    return info != nullptr ? info->Is_Grantable() : false;
+}
+
+// zh: 0x00488CE0 wb: 0x007277AF
+void ScienceStore::Get_Name_And_Description(ScienceType science, Utf16String &name, Utf16String &description)
+{
+    auto *info = Get_Science_Info(science);
+    if (info == nullptr) {
+        return;
+    }
+    name = info->Get_Name();
+    description = info->Get_Description();
+}
+
+// zh: 0x00488D50 wb: 0x007277F7
+bool ScienceStore::Player_Has_Prereqs_For_Science(Player *player, ScienceType science)
+{
+    // TODO: Requires further implementation of Player
+#ifdef GAME_DLL
+    return Call_Method<bool, ScienceStore, Player *, ScienceType>(
+        PICK_ADDRESS(0x00488D50, 0x007277F7), this, player, science);
+#else
+    return false;
+#endif
+}
+
+// zh: 0x00488DC0 wb: 0x00727867
+bool ScienceStore::Player_Has_Root_Prereqs_For_Science(Player *player, ScienceType science)
+{
+    // TODO: Requires further implementation of Player
+#ifdef GAME_DLL
+    return Call_Method<bool, ScienceStore, Player *, ScienceType>(
+        PICK_ADDRESS(0x00488DC0, 0x00727867), this, player, science);
+#else
+    return false;
+#endif
+}
+
+// zh: 0x00488950 wb: 0x0072752B
+void ScienceStore::Parse_Science_Definition(INI *ini)
+{
+    const auto *name = ini->Get_Next_Token();
```

Comment:
What is `Get_Next_Token`? Does it return a `const char*` or `Utf8String`? Looks suspicious with the auto.

Reply:
Doesn't really matter what it returns and now it's all ready for if the type is modified.

Reply:
It looks like bug if `ini->Get_Next_Token()` were to return Utf8String by value. From review I can't see what it returns, so I am raising this here.

Reply:
The beauty of auto* is that it would be a compile error if it returned it by value and this would have been subtly broken if i had used char* and it turned out the function returned a utf8string by value. See https://godbolt.org/z/v88Tdc33f where I have a mock class "test" that is similar to utf8string with the operator that converts to char*.

Reply:
Interesting. I woud have expected the implicit cast to take effect there as well. On the other hand, it probably makes most sense the way it is, because there could be multiple cast operator overloads, and there would be no way auto could decide on that either.